activemq-client/src/main/java/org/apache/activemq/ActiveMQSession.java

```diff
@@ -2049,6 +2066,146 @@ protected void send(ActiveMQMessageProducer producer, ActiveMQDestination destin
+
+            // Set the message id.
+            if (msg != message) {
+                message.setJMSMessageID(msg.getMessageId().toString());
+                // Make sure the JMS destination is set on the foreign messages too.
+                message.setJMSDestination(destination);
+            }
+            //clear the brokerPath in case we are re-sending this message
+            msg.setBrokerPath(null);
+
+            msg.setTransactionId(txid);
+            if (connection.isCopyMessageOnSend()) {
+                msg = (ActiveMQMessage)msg.copy();
+            }
+            msg.setConnection(connection);
+            msg.onSend();
+            msg.setProducerId(msg.getMessageId().getProducerId());
+            if (LOG.isTraceEnabled()) {
+                LOG.trace(getSessionId() + " sending message: " + msg);
+            }
+            if (completionListener==null && sendTimeout <= 0 && !msg.isResponseRequired() && !connection.isAlwaysSyncSend() && (!msg.isPersistent() || connection.isUseAsyncSend() || txid != null)) {
+                this.connection.asyncSendPacket(msg);
+                if (producerWindow != null) {
+                    // Since we defer lots of the marshaling till we hit the
+                    // wire, this might not
+                    // provide and accurate size. We may change over to doing
+                    // more aggressive marshaling,
+                    // to get more accurate sizes.. this is more important once
+                    // users start using producer window
+                    // flow control.
+                    int size = msg.getSize();
+                    producerWindow.increaseUsage(size);
+                }
+            } else {
+                if (sendTimeout > 0 && completionListener==null) {
+                    this.connection.syncSendPacket(msg, sendTimeout);
+                }else {
+                    CompletionListener wrapperCompletionListener = null;
+                    if (completionListener != null) {
+                        // Make the Message object unaccessible and unmutable per Jakarta Messaging 3.1 spec section 7.3.9 and 7.3.6
+                        msg.setIsMessageAccessibleAndMutable(false);
+                        numIncompletedAsyncSend.doIncrement();
+                        wrapperCompletionListener = new CompletionListener() {
+                            @Override
+                            public void onCompletion(Message message) {
+                                try {
+                                    if (message instanceof ActiveMQMessage) {
+                                        ((ActiveMQMessage) message).setIsMessageAccessibleAndMutable(true);
+                                    }
+                                    isInCompletionListenerCallback.set(true);
+                                    isProducerInCompletionListenerCallback.set(true);
```

Comment:
```
isInCompletionListenerCallback.set(true);
isProducerInCompletionListenerCallback.set(true);
```
Why are these methods invoked on completion? I would have expected them to have been invoked at a similar time as `numIncompletedAsyncSend.doIncrement();`. Maybe I am missing something?

Reply:
Because when the the counter increment by 1, the completionListener is not triggered yet. So you want to increment the number of incompleted async send but don't want to set the ThreadLocal object. Also if you do it outside of the completionListener#onComplete method, it will set at the application thread, not the thread that is executing the onComplete function.

Reply:
That makes sense, thanks for the explanation.